File: docs/stack/1944_hard/visible_people.cpp
#include <vector>
#include <stack>
using namespace std;

vector<int> countVisibilities(vector<int> &heights) // LeetCode Q.1944.
{
    stack<int> stack; // A stack of heights.

    vector<int> visibilities(heights.size(), 0);

    for (int idx = heights.size() - 1; idx >= 0; idx--)
    {
        int height = heights[idx];

        while (!stack.empty() && stack.top() < height)
        {
            stack.pop();
            visibilities[idx]++;
        }

        if (!stack.empty())
        {
            visibilities[idx]++;

            if (stack.top() == height)
                stack.pop();
        }

        stack.push(height);
    }

    return visibilities;
}

File: docs/stack/1944_hard/visible_people.py

def count_visibilities(heights: list[int]) -> list[int]:  # LeetCode Q.1944.
    stack = []
    visibilities = [0] * len(heights)

    for idx in range(len(heights) - 1, -1, -1):  # Backward iteration.
        height = heights[idx]
        
        while stack and stack[-1] < height:  # People on my left side can't see them.
            stack.pop(-1)
            visibilities[idx] += 1  # I can see unblocked shorter people.

        if stack:  # I can always see my next non-shorter person (if exists).
            visibilities[idx] += 1

            if stack[-1] == height:  # People on my left side can't see this person.
                stack.pop(-1)
        
        stack.append(height)

    return visibilities
